astropy/table/operations.py

```diff
@@ -129,6 +130,98 @@ def join(left, right, keys=None, join_type='inner',
     return out
 
 
+def setdiff(table1, table2, keys=None):
+    """
+    Take a set difference of table rows.
+
+    The row set difference will contain all rows in ``table1`` that are not
+    present in ``table2``. If the keys parameter is not defined, all columns in
+    ``table1`` will be included in the output table.
+
+    Parameters
+    ----------
+    table1 : Table object
+        ``table1`` is on the left side of the set difference.
+    table2 : Table object
+        ``table2`` is on the right side of the set difference.
+    keys : str or list of str
+        Name(s) of column(s) used to match rows of left and right tables.
+        Default is to use all columns in ``table1``.
+
+    Returns
+    -------
+    diff_table : `~astropy.table.Table` object-
+        New table containing the set difference between tables. If the set
+        difference is none, an empty table will be returned.
+
+    Examples
+    --------
+    To get a set difference between two tables::
+
+      >>> from astropy.table import setdiff, Table
+      >>> t1 = Table({'a': [1, 4, 9], 'b': ['c', 'd', 'f']}, names=('a', 'b'))
+      >>> t2 = Table({'a': [1, 5, 9], 'b': ['c', 'b', 'f']}, names=('a', 'b'))
+      >>> print(t1)
+       a   b
+      --- ---
+        1   c
+        4   d
+        9   f
+      >>> print(t2)
+       a   b
+      --- ---
+        1   c
+        5   b
+        9   f
+      >>> print(setdiff(t1, t2))
+       a   b
+      --- ---
+        4   d
+
+      >>> print(setdiff(t2, t1))
+       a   b
+      --- ---
+        5   b
+    """
+    if keys is None:
+        keys = table1.colnames
+
+    #Check that all keys are in table1 and table2
+    for tbl, tbl_str in ((table1,'table1'), (table2,'table2')):
+        diff_keys = np.setdiff1d(keys, tbl.colnames)
+        if len(diff_keys) != 0:
+            raise ValueError("The {} columns are missing from {}, cannot take "
+                             "a set difference.".format(diff_keys, tbl_str))
+
+    # Make a light internal copy of both tables
+    t1 = table1.copy(copy_data=False)
+    t1.meta = {}
+    t1.keep_columns(keys)
+    t1['__index1__'] = np.arange(len(table1))  # Keep track of rows indices
+
+    t2 = table2.copy(copy_data=False)
+    t2.meta = {}
+    t2.keep_columns(keys)
+    t2['__index2__'] = np.zeros(len(table2), dtype=np.uint8)  # Dummy column
+
+    t12 = _join(t1, t2, join_type='left', keys=keys,
+                metadata_conflicts='silent')
```

Comment:
Should metadata_conflicts be a parameter as for other helper functions?

Reply:
No. The metadata of the right table `t2` is ignored.  I think for the stated purpose of returning rows from the left table that are not in the right table, this makes sense.  Conceptually (and in code) it is simply selecting appropriate rows of the left table, so metadata join is not desired.